File: BOJ(Java 공부용)/1927.java
import java.io.*;
import java.util.PriorityQueue;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().split(" ")[0]);
        PriorityQueue<Integer> q = new PriorityQueue<>();
        for(int i=0;i<n;i++) {
            int num = Integer.parseInt(br.readLine().split(" ")[0]);
            if(num == 0) {
                if(!q.isEmpty()) bw.write(q.poll()+"\n");
                else bw.write("0\n");
            }
            else {
                q.add(num);
            }
        }
        bw.flush();
        bw.close();
    }
}

File: BOJ(Kotlin)/1927.kt
import com.sun.org.apache.xpath.internal.operations.Bool
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main(args : Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val n = br.readLine().toInt()
    var pq = PriorityQueue<Int>()
    for (i in 0 until n) {
        val num = br.readLine().toInt()
        when {
            num == 0 -> {
                if (!pq.isEmpty()) bw.write(pq.poll().toString()+"\n")
                else bw.write("0\n")
            }
            else -> {
                pq.add(num)
            }
        }
    }
    bw.flush()
    bw.close()
}